Zend/zend_inheritance.c

```diff
@@ -173,72 +174,205 @@ static zend_always_inline zend_bool zend_iterable_compatibility_check(zend_arg_i
+/* caller is responsible for adding any necessary refcount */
+static zend_string *_get_parent_class_name(zend_class_entry *ce)
+{ /* {{{ */
+	zend_string *pname;
+
+	if (ce->ce_flags & ZEND_ACC_LINKED) {
+		ZEND_ASSERT(ce->parent);
+		pname = ce->parent->name;
+	} else {
+		pname = ce->parent_name;
+	}
+
+	/* If there is a parent, it must have a name */
+	ZEND_ASSERT(pname);
+	return pname;
+}/* }}} */
+
+static
+zend_string *_resolve_parent_and_self(const zend_function *fe, zend_string *name)
+{ /* {{{ */
+	zend_class_entry *ce = fe->common.scope;
+	/* if there isn't a class then we shouldn't be resolving parent and self */
+	ZEND_ASSERT(fe->common.scope);
+
+	switch (zend_get_class_fetch_type(name)) {
+		case ZEND_FETCH_CLASS_PARENT:
+			name = _get_parent_class_name(ce);
+			break;
+
+		case ZEND_FETCH_CLASS_SELF:
+			name = fe->common.scope->name;
+			break;
+
+		case ZEND_FETCH_CLASS_DEFAULT:
+			/* already resolved */
+			break;
+
+		case ZEND_FETCH_CLASS_STATIC:
+			/* Note: this currently a syntax error */
+			zend_error_noreturn(E_COMPILE_ERROR,
+				"Cannot use static as a type constraint in %s::%s()",
+				ZEND_FN_SCOPE_NAME(fe), ZSTR_VAL(fe->common.function_name));
+
+		EMPTY_SWITCH_DEFAULT_CASE();
+	}
+
+	return zend_string_copy(name);
+} /* }}} */
+
+typedef enum {
+	CONTRAVARIANT = -1,
+	INVARIANT = 0,
+	COVARIANT = +1,
+} _variance;
+
+static
+int _check_zce_variance(
+	const zend_class_entry *ce1, const zend_class_entry *ce2,
+	_variance variance) /* {{{ */
 {
-	ZEND_ASSERT(ZEND_TYPE_IS_SET(fe_arg_info->type) && ZEND_TYPE_IS_SET(proto_arg_info->type));
+	if (ce1 == ce2) return 1;
 
-	if (ZEND_TYPE_IS_CLASS(fe_arg_info->type) && ZEND_TYPE_IS_CLASS(proto_arg_info->type)) {
-		zend_string *fe_class_name, *proto_class_name;
-		const char *class_name;
-		size_t class_name_len;
+	switch (variance) {
+		case INVARIANT:
+			// todo: does ce1 == ce2 above catch all possible cases?
+			return 0;
+		case COVARIANT:
+			return instanceof_function(ce1, ce2);
+		case CONTRAVARIANT:
+			return instanceof_function(ce2, ce1);
 
-		fe_class_name = ZEND_TYPE_NAME(fe_arg_info->type);
-		class_name = ZSTR_VAL(fe_class_name);
-		class_name_len = ZSTR_LEN(fe_class_name);
-		if (class_name_len == sizeof("parent")-1 && !strcasecmp(class_name, "parent") && proto->common.scope) {
-			fe_class_name = zend_string_copy(proto->common.scope->name);
-		} else if (class_name_len == sizeof("self")-1 && !strcasecmp(class_name, "self") && fe->common.scope) {
-			fe_class_name = zend_string_copy(fe->common.scope->name);
-		} else {
-			zend_string_addref(fe_class_name);
-		}
+		EMPTY_SWITCH_DEFAULT_CASE();
+	}
+}
+
+/* This int return is not a boolean, but will often be treated this way.
+ *   *  0 means there is definitely an error.
+ *   *  1 means there are definitely not any errors.
+ *   * -1 means there are no known errors but it is not known to be good
+ *     either (there is not enough information at the moment).
+ */
```

Comment:
Maybe make an enum for this? At least so the return values inside the function are clear.

Reply:
How about this for naming?

```c
typedef enum {
	INHERITANCE_INDETERMINANT = -1,
	INHERITANCE_ERROR = 0,
	INHERITANCE_SUCCESS = 1,
} _inheritance_status;
```

Reply:
Looks reasonable. s/INDETERMINANT/UNKNOWN maybe ^^